### PR TITLE
configs: validate: add validation for bind-mount fsflags

### PR DESCRIPTION
**Pull request by cyphar**
Bind-mounts cannot have any filesystem-specific "data" arguments, because the kernel ignores the data argument for MS_BIND and MS_BIND|MS_REMOUNT and we cannot safely try to override the flags because those would affect mounts on the host (these flags affect the superblock).

It should be noted that there are cases where the filesystem-specified flags will also be ignored for non-bind-mounts but those are kernel quirks and there's no real way for us to work around them. And users wouldn't get any real benefit from us adding guardrails to existing kernel behaviour.

See the discussion in opencontainers/runtime-spec#1222. I am a little worried this could break users, but I also would be very surprised if users were specifying flags to bind mounts that are always ignored.

Signed-off-by: Aleksa Sarai <cyphar@cyphar.com>

### CODE REVIEW

File: libcontainer/configs/validate/validator.go
Comment (kolyshkin) [inline libcontainer/configs/validate/validator.go:291]: nit: `errors.New`

**Review by kolyshkin — APPROVED** [review]
lgtm

@rata ptal

**Comment by rata** [comment]
The code per-se is fine. But I'm not sure I understand is the intent of this PR. In the description is said quite clearly too:

> I am a little worried this could break users, but I also would be very surprised if users were specifying flags to bind mounts that are always ignored.

If this always worked, why do we want to start throwing an error? What would be simpler for us in the future if we do this?

I'd like to think a warning can help as a first step if we want to go on this direction, but no one really looks at the warnings we throw at this part of the stack today, so I don't think it will help :-/

**Comment by cyphar** [comment]
At the moment we are ignoring flags the user has explicitly requested and there is no way for us to not ignore them (you by defintion cannot set fs-specific flags on bind-mounts because they are per-superblock). I see this as being related to the work in #3967. The runtime-spec has frustratingly loose wording on how mounts should be handled in these sorts of edge cases.

Perhaps it would be less aggressive to make this a warning and we can return an error in 1.3 if nobody complains.

**Comment by rata** [comment]
> Perhaps it would be less aggressive to make this a warning and we can return an error in 1.3 if nobody complains.

I think we already have a bunch of changes in 1.2, I prefer to not add more possibly breaking changes.

But what do we have to win by returning an error? Will something be better for us, or it just has the potential to break users? Can't we even assign data to `""` and throw a warning and consider it done?

File: libcontainer/configs/validate/validator.go
Comment (rata) [inline libcontainer/configs/validate/validator.go:291]: What does filesystem-specific mean here? In the runtime spec I don't see any differentiation between fs-specific and not specific flags (https://github.com/opencontainers/runtime-spec/blob/main/config.md#linux-mount-options). Am I missing something?

It will be nice to throw a clearer error for the user here. **The` idmap` and `ridmap` flags have just been added to that table and those will be valid, while the rest not**. It will be nice if we can throw a better error or separate the flags in the spec, so a user can know what is a valid combination and what it is not.

**Review by rata — COMMENTED** [review]
Ok, I think the value this adds is that we will reject flags we don't recognize.

I think this is nice (would be even better to do it in crun and reflect this in the spec in some way), and I'm up for backporting this to 1.1 so we find out early if this causes issues to users.

Left some small comments on the error and minor nits.

File: libcontainer/configs/validate/validator_test.go
Comment (kolyshkin) [reply libcontainer/configs/validate/validator_test.go:433]: I think it's OK to have it explicitly set, for clarity. Not a big deal either way.

File: libcontainer/configs/validate/validator.go
Comment (kolyshkin) [reply libcontainer/configs/validate/validator.go:291]: In fact, runtime spec mentions both "filesystem-specific" and "filesystem-independent" mount options, referring to `mount(8)` (and the only "filesystemd-specific" options it mentions are mount propagation modes for bind mounts).

I guess we need to change the wording in the spec to distinguish the real filesystem-specific mount options (not documented in runtime-spec at all) from the bind mount propagation options. WDYT @AkihiroSuda ?

Comment (AkihiroSuda) [reply libcontainer/configs/validate/validator.go:291]: > I guess we need to change the wording in the spec to distinguish the real filesystem-specific mount options (not documented in runtime-spec at all) from the bind mount propagation options. WDYT @AkihiroSuda ?

SGTM, PR is wanted

Comment (kolyshkin) [reply libcontainer/configs/validate/validator.go:291]: > I guess we need to change the wording in the spec to distinguish the real filesystem-specific mount options (not documented in runtime-spec at all) from the bind mount propagation options. 

Opened https://github.com/opencontainers/runtime-spec/pull/1228

Comment (kolyshkin) [reply libcontainer/configs/validate/validator.go:291]: @rata as for this PR, I think the error message wording ("bind mounts cannot have any filesystem-specific options applied") is totally correct.

**Review by kolyshkin — APPROVED** [review]
Couple of nits but overall LGTM.

File: libcontainer/configs/validate/validator.go
Comment (rata) [reply libcontainer/configs/validate/validator.go:291]: @kolyshkin thanks for the PR! SGTM this error, then  :)

**Review by rata — APPROVED** [review]
Overall LGTM, no need to fix the nits in this PR, and if you fix them here no need to re-review, they are trivial

**Comment by lifubang** [comment]
> I am a little worried this could break users

Is this need to be backported to '1.1'? I am a bit worried too.

> Perhaps it would be less aggressive to make this a warning and we can return an error in 1.3 if nobody complains.

If we change to output a warning msg, I think we can backport it to '1.1'.

**Comment by kolyshkin** [comment]
> > I am a little worried this could break users
> 
> Is this need to be backported to '1.1'? I am a bit worried too.

My thinking is, it will help users to find problems in their setups before we release 1.2.0 (meaning, instead of piling everything to 1.2.0, we can use the 1.1.10 opportunity to introduce this change -- and, if it is really breaking users setups, revert it).

A warning seems better, the only problem is, no one sees these warnings, because no one uses runc directly, and (AFAIK) upper level tools do not show runc output to users.

OTOH I don't have a stong opinion, and since this is definitely not a bug fix, let's not backport it.

File: libcontainer/configs/validate/validator_test.go
Comment (cyphar) [reply libcontainer/configs/validate/validator_test.go:433]: Even purely for cosmetic reasons, I prefer to have it be explicit.

**Comment by cyphar** [comment]
Fixed up the nits.

I agree with @kolyshkin that it would be nice if warnings were useful, but they're not so making it warning-only in 1.2 (or backporting a warning-only version to 1.1) are of limited usefulness...

**Comment by kolyshkin** [comment]
Also, in case it causes a major breakage, we can address that in a 1.2.z stable release in a timely manner.

**Comment by kolyshkin** [comment]
@opencontainers/runc-maintainers PTAL

File: libcontainer/configs/validate/validator.go
Comment (AkihiroSuda) [inline libcontainer/configs/validate/validator.go:293]: Can be just a warning?

**Comment by thaJeztah** [comment]
> because the kernel will completely ignore these flags

This PR seems reasonable to me, although we must be careful to what extent it's "our" responsibility to validate things (or to leave it to the kernel). I still recall at some point we added validation for "valid environment variable names", which turned out to be a bit over-enthusiastic, as there were many systems _depending_ on names outside of the "safe" range.
So for situations where it doesn't "really" complicate things for us, I'm leaning towards leaving it to the kernel to decide whether something should be considered valid or not.

File: libcontainer/configs/validate/validator.go
Comment (cyphar) [reply libcontainer/configs/validate/validator.go:293]: The problem is that (at least, according to some other recent issues) most people can't see runc warnings because higher-level runtimes don't provide them to users, so adding a warning doesn't do much. I think we had the same issue with the relative-mount-paths validation.

**Comment by cyphar** [comment]
@thaJeztah The problem is that in this case (unlike the variable name cases, and most other cases where we don't restrict things the kernel allows), not checking means we are ignoring any set flags on bind-mounts, which means that the `idmap` and `ridmap` flags won't produce errors on older runc versions -- meaning that adding new mount options in a backwards-compatible way (where users are told the flag is being ignored) isn't possible for bind-mounts without this kind of validation.

The fact the kernel ignores the data argument for `MS_BIND` is due to the way the `mount(2)` syscall was designed (it's a multiplexer where some arguments are not relevant for some operations, and has finally been broken up into a much better design with the new mount API) and it's not possible for any future Linux kernel version to change the behaviour.

File: libcontainer/configs/validate/validator.go
Comment (cyphar) [reply libcontainer/configs/validate/validator.go:293]: FWIW, I would prefer if we made this a warning for one release and then upgraded it to an error in runc 1.3, but if warnings are invisible to most users then there's not much point.

**Review by AkihiroSuda — APPROVED** [review]

**Pull request merged**
